### DISCUSSION TITLE
Storybook tags, filter UI, and badges

### DISCUSSION THREAD

**Discussion by shilman** [storybookjs/storybook#30091] [Show and tell]
In Storybook 8.4 we released [Storybook tags with interactive filter UI, plus an addon for badges in the sidebar](https://storybook.js.org/blog/storybook-tags/). ([docs](https://storybook.js.org/docs/writing-stories/tags), [addon](https://github.com/Sidnioulz/storybook-addon-tag-badges)).

Curious if you're finding tags useful, whether you have requests, and if you have use cases that you can share with the community!

![image](https://github.com/user-attachments/assets/b66c5a06-bf32-413a-9b4d-696bf1d137b2)

**Comment by smlimasu** [comment]
Thank you for your work. 

I would like to have a regex exclude filter option for tags in the sidenav Tag dropdown menu, because we have internal tags that we don't want to show.

Also, it would be great to have a tag like "test-only" or "internal" that displays stories in the sidenav only when they are checked (so that testers can toggle them on, but normal users don't in default. 

Other use cases: I have made a docs component that shows various tags above the docs title as chips, also some tag-chips above each story-title in the docs. It reads from the story/meta. Would be great if there was a storybook block or option in stories-block, maybe with the tags addon (great work, frog!).

For the Controls/Argtypes tables, I enhance my Argtypes descriptions with various jsdoc tags and their info (filling the missing gap from compodoc.json) like deprecated, since (version XY), param. I display them in different colors and styles as chips to give my users more info about the argtype. Would be great if a) you could specify in preview which tags to include (like exclude internal, ignore (I know, this can be done via compodoc build options), include deprecated or version, and we would have these jsdoc info generated the same for all frameworks and b) set global styles to these tags like in the tags addon, maybe the same like create theme options. 

Wish you happy holidays!

**Reply by shilman** [reply]
Thanks so much for the wonderful feedback! The "force `test` tags to show" UI is a great idea and addresses a problem that I've heard from many different users. I don't know the best way to implement that yet, but I'll think about it and will share a short RFC here once I have a proposal.

Would also be really cool to support JSdoc tags somehow. Though probably a non-starter based on the requirement that tags be statically analyzed during indexing.

@Sidnioulz I wonder whether the tags doc block could be a good addition to the Tags Badges addon.

**Reply by kylegach** [reply]
Sorry for being slightly off-topic here, but your Storybook sounds _amazing_, @smlimasu! 😍

**Reply by Sidnioulz** [reply]
@smlimasu please upvote this issue :) I'll likely work on it in a month-ish.

https://github.com/Sidnioulz/storybook-addon-tag-badges/issues/30

**Reply by Sidnioulz** [reply]
> Also, it would be great to have a tag like "test-only" or "internal" that displays stories in the sidenav only when they are checked (so that testers can toggle them on, but normal users don't in default.

I can see a few ways to achieve this.

There could be a default filter. If you wanted that, should it _just_ be a default filter being applied (and you can disable it in the UI), or should the UI display an "empty"/"initial" state despite these filters being applied?

If the former, you could (soon?) use query parameters to pass URLs that configure those default filters. With a bit of redirection magic, you could enforce the default tags you want without having to explicitly set the query parameters.

---

If the latter, it would complicate the UI a bit, and you'd only be able to enforce one default rather than several UI presets for different audiences.

Should it go further? Should you have a general setting that applies content filters / default values for other filters _on top of_ the tag filter? And then a select in the toolbar or sidebar to choose a preset?

Should such a preset be available in MDX so you could add `<DevOnly>`, `<QAOnly>`, etc. blocks in your MDX doc?

I suspect most of that can be achieved right now with an addon. The only difficult part would be negotiating what happens with the Storybook tag filter UI when you apply a preset.

---

It could also be a permanent filter that filters out which stories get included in the SB instance, and then you deploy different versions for general documentation and for testers.

In which case, would it differ from instrumenting your stories regex and separating tester/documentation stories? This is already achievable but maybe it's not clear how it can be done. It does involve a more complex deployment process, though.

**Reply by smlimasu** [reply]
@Sidnioulz these are important, elaborate thoughts and use cases you sketched here. Thanks. Initially, I just thought about it as a kind of reverse filter Tag in the tags menu where selecting a predefined Tag like "test-only" would add more content to the sidebar.

I know I can deploy different versions but I thought there might be a more convenient method.

The scenario to have different presets, just like an environment preset, but configurable by the user is an idea I like a lot and I would use that. 

I imagine it perhaps like the version switcher button you have with refs, or some people have build as an addon. Or the theme switcher globals. When there was a global button, maybe in the toolbar or in the sidenav or sidenav config, that you could easily configure or add options in the manager based on globals, tags, filters, parameters, themes, languages or versions to serve different stakeholders needs that would be awesome. I think it's already possible but you would need to write your own addon/advanced globals for it. For some weeks, because I did not like some of those toggles being in the toolbar, I have inserted a second top nav bar like in the storybook docs to let users easily toggle different presets/modes/sections but nah, felt wrong. Isn't there this experimental sidenav_top for add-ons? I hesitated to use it atm and to be honest, I found the documentation for local add-ons is not extensive enough to just easily build some of these features without a headache. 

So yeah, making it more easily for devs to address some of these use cases by providing more build in features/working addons without needing to deep dive into storybook mechanics and reinventing the wheel would be so great, I would try to contribute. But what and how exactly, that depends what other users need and what's feasible in the storybook teams pov and goals.

**Reply by smlimasu** [reply]
@shilman not sure if we mean the same with the jsdocs tags. I understand that it's difficult to tie jsdoc tags to e.g. the tags-filter-menu, but thats not what i thought of.

My goal is another. Since jsdoc tags are an essential part of code documentation, written extensively in the components source code, it's important that storybook shows that info where it could. Since we're talking about tags, for me they are different kind, but we should talk about their current state and their roadmap too, i think. 

This comment only relates to the enhancement of the static text display of various jsdoc / argtype info in the argtype-block/controls-table. I hope I understand correctly that this is no problem with the indexer? 

Here is an example what I mean, left my custom implementation, right the (angular SB) default behaviour: 

![jsdoctags](https://github.com/user-attachments/assets/f1b16a98-21c6-4c2b-8c34-b6c144061056)

There are different feature requests and open (or closed, but not working properly atm) bugs around displaying docgen jsdoc information or jsdoc tags in the argstable. See f.ex. #9721, #28830 , #26154 , #21794 . Looking at the source code, I see the various frameworks do not  extract a standardized amount of info (if available trough their docgen api ofc). Example: In Angular and Svelte, deprecated is currently not displayed at all, but in react it is. There are other differences with see, param, since, too. 

It would be so great if there was a consensus/goal or public guidance by the maintainers for contributors on what needs to be extracted from the jsdoc, if it should be joined to the description or if it should go into that array of tags, and how to display it (what order, what style/color, which column, striketrough the argtype.name), etc. in the table? So that contributors can all aim to implement the desired outcome step by step.

I'd really like to contribute to this, and fix the problems especially for angular, or implement features like in my screenshot, but for example I'm not sure if it would be okay to add colors or styleclasses to the tags or not. Should I only fix "deprecated" or go on and show more relevant jsdoc tags? Who or where to ask? What do other users think?

Just some food for thought for further discussions, thank you for reading ( :

**Reply by shilman** [reply]
Hi @smlimasu, thanks for clarifying.

Indeed, I thought you were referring to Storybook's notion of tags, which have the hard requirement of static analysis in the indexer. It's technically possible to do add Jsdoc parsing to the static analysis, but it's a heavy lift.

The information in the ArgsTable is a different code path. I would be open to a technical discussion on how we might go about improving that. Perhaps we could work on a plan together?

**Reply by Sidnioulz** [reply]
Honestly I can see the appeal to having tags for argTypes. It is _so_ common for library maintainers to want to highlight when a prop was introduced or deprecated; having built-in tags support (e.g. a `tags` property on the `argTypes` item and then addons can display those if there's the right ArgsTable hook?) would be amazing.

Regarding the earlier discussion on default filters / profiles / etc, I think the easiest resolution for a v1 would be to provide API endpoints to set the tag filter, reachable by addons. Then anyone is free to build default filters with an API call in their manager, or an addon for profile switching. WDYT @shilman?

**Reply by shilman** [reply]
@Sidnioulz there's already a sidebar filtering API in the manager. I believe it's intentionally undocumented and prefixed with `experimental_` but it's there. That API is a filter function so it's quite flexible. The tag filter UI uses that API under the hood.

If you want manager API's to set the tag filter specifically, that should also be possible. I'm open to it, but I'd only want to increase the surface area in that way if the use cases were better articulated.

**Reply by Sidnioulz** [reply]
I think it would provide a better experience to end users if there was an API that lets addon authors configure tag filters, so that what they do is reflected in the UI rather than hidden, but that's just an opinion and I thoroughly agree with you that we need more fleshed out use cases and more requests before adding maintenance surface.

**Comment by Sidnioulz** [comment]
Making a dedicated thread to discuss `tags` in `argTypes`.

### Use cases

Brought up by @smlimasu but also something I've experienced as a library maintainer.

* Deprecated props
* Prop versioning

### Opportunities

Having built-in tags support would allow showing badges in the ArgsTable next to props, and possibly also in the controls description, though likely with extra development effort or additional work in stories files.

### Challenges

Since `argTypes` are not statically loaded, the tag filter UI cannot support such tags.

**Reply by smlimasu** [reply]
To add to the list of use cases: 

- param tag for methods with param description
- returns tag with description
- proper formatting for see, link and example
- make sure to have option to hide "internal" or "ignore" or "private/protected" props (like in the compodoc options for angular), maybe as an option in the preview.controls object 

- should all tags and their description/comment part go into an array "jsdoc tags" of the argtype object to allow users implement custom things? I would say yes, this way users could also add additional tags and descriptions in the argtypes of the story manually e.g. when the argtype is not pulled from the prop jsdoc. 

styling decisions:
- style deprecated in an "orange" color (should this be configurable by the user? Then add tags styling to the theme options array or at least add a style class to each tags span)
- should deprecated args be sorted last in the argstable, with strike trough and lighter text color? See mock from 2021.
- style since/version in a badge color style (maybe primary color?)
- order of tags, I would suggest deprecated first, then the overall arg description, then "versioning", then param/returns and the default value.

@shilman yes I am open to help to discuss or mock a plan/draft for this, thank you.

**Reply by Sidnioulz** [reply]
As tags would be simple word lists, they couldn't substitute the way function arg descriptions are done in the ArgsTable.

I'm curious if addon hooks can be provided in the ArgsTable, @shilman? Is there a way we can avoid increasing the code footprint and allow addon authors to do more? Going back on the idea of custom controls (there's a PR for that somewhere in the repo) and my unexplored idea of JSX controls, providing an entry point into the table for addons could have extra benefits.

**Reply by smlimasu** [reply]
@Sidnioulz yes, but they have a tagname and description, like there already is the jsdoctags type in the argtype in the codebase, see this story https://github.com/storybookjs/storybook/blob/2759c5bf3aacefbb4b0095fabb69d9a500a482aa/code/lib/blocks/src/components/ArgsTable/ArgRow.stories.tsx#L291

https://github.com/storybookjs/storybook/blob/1664776b5421990ebee22ecc3ab3e0d7c0796e21/code/renderers/react/template/stories/docgen-components/jsdoc/argTypes.snapshot#L10

when i set the jsdoc tags in the argTypes of my story like this (if needed, for simpler use cases, not as a substitution, e.g. when you need to define argTypes for documentation purposes yourself in some cases)

``` 
const meta = {
  argTypes: {
      backgroundColor: {
        description: 'sets color of the background',
        control: 'color',
        table: {
          defaultValue: {
            summary: '#F5F5F5',
          },
          jsDocTags: {
            deprecated: 'this arg is deprecated',
          },
        },
      },
    },
} satisfies Meta<typeof Button>;
  ```
  
  it currently shows the deprecated "tag" and the deprecated string message in SB react stackblitz. So why don't we add this behaviour to the argType type's documentation/typespec officially.
  
Maybe we mean something different here, i am talking about concatenating all tags, their descriptions and the prop description, param, returns etc in the end to one "description text"-cell in the argstable, like I did it in my screenshot, you are talking about adding the jsdoc tags in an extra column or something? Could you clarify? 

yes, addon hooks would be great.

**Reply by Sidnioulz** [reply]
> Maybe we mean something different here, i am talking about concatenating all tags, their descriptions and the prop description, param, returns etc in the end to one "description text"-cell in the argstable, like I did it in my screenshot, you are talking about adding the jsdoc tags in an extra column or something? Could you clarify?

We do :)
* JSDoc tags refer to https://jsdoc.app/; in Storybook there are some JSDoc tags in `argTypes[...].table.jsDocTags`
* Storybook tags are a Storybook-only list of words applied to stories, components meta, docs entries (and by extrapolation, to all Storybook Hash entries)

I refer to the latter, not the former. Having tags on argTypes could extend beyond the `table`: I could possibly also want to print a summary of new props or deprecated props in a MDX block that I add to my autodocs template, for instance. I could also exploit that information in a future Figma CodeConnect integration.

**Comment by Grubsniff** [comment]
Thank you for this valuable contribution! The tagging system is particularly useful for organizing stories. 

However, I'd like to expand on @smlimasu feedback regarding story visibility control.

Current Challenge:
We're running a production Storybook instance where test stories are visible alongside production components. With hundreds of test stories, this creates unnecessary noise for our end users.

Feature Request:
Could we implement a default visibility filter based on tags? 

- Ability to automatically hide stories marked with a 'x' tag
- Option to toggle visibility of filtered stories when needed
- Default hide/show preferences that persist across sessions

This would significantly improve the user experience for teams using Storybook in production environments while maintaining the ability to access test stories when needed.

Has there been any consideration for implementing this kind of filtering system? Happy to provide more context about our use case if helpful.

Thanks again for maintaining this project!

**Reply by shilman** [reply]
I think this is the logical next step for tags! I scoped things down to keep the first version small, but def have this in mind.

For tests specifically we have [a slightly different idea](https://github.com/storybookjs/storybook/discussions/30119) for differentiating stories from tests, but I think this request still holds for any kind of story that clutters the sidebar.

**Reply by Grubsniff** [reply]
Thanks for the reply. 

This is great to hear that it is in scope as an enhancement. 

Thanks for sharing the tests approach, that actually seems really valuable for one of our other storybook teams!

Has this idea been implemented as of yet? or would we expect that in a future version?

**Reply by shilman** [reply]
@Grubsniff that's great! The `.test` proposal is still just at concept stage but will very likely be ready late this spring for 9.0.

The tags improvements are not on the roadmap yet but are on my radar. Will try to squeeze them in as they are a smaller lift. Will post back here when there's progress!

**Comment by akibmayadav** [comment]
Thank you for the amazing work! 
Would be great to have this `URL parameters for tag filters so that you can share a filtered view of your Storybook`. Is there any way to use url parameters to share specific views at this point of time ? Or is it possible to set some default selections of the filter ?

**Reply by ptbrowne** [reply]
It would be great to be able to have a link to a pre-filtered view. This could help sharing a link to components needing a visual review, or that are new.

**Reply by shilman** [reply]
Good one. Any thoughts on how that might look? A CLI option like `storybook build --includeTags new --excludeTags experimental`?

**Comment by enemyXXX** [comment]
That really nice, but what about filtering by tags using search input? I guess it will be really helpful for huge projects with thousand stories, where each story has kind of shortcut with often tries to search it

**Reply by shilman** [reply]
We don't have a search-based UI yet. However, if you filter by tag using the UI, it will also filter down the search results AFAIK.

**Comment by oriooctopus** [comment]
Could this be used to SEO stories? For example, if I have a component called `FinalStep`, I'd like it to come up if somebody searches `DoneStep`. It would be nice if tagging allowed for this

**Comment by breakone** [comment]
Is there a possibility to show tags, e.g. the deprecated, in the components tree? Something like this?
![image](https://github.com/user-attachments/assets/d143eb38-08eb-469b-9a7d-8a270a158cc0)

**Reply by Sidnioulz** [reply]
Yes, use the tag badges addon: Https://github.com/sidnioulz/storybook-addon-tag-badges

**Reply by breakone** [reply]
🙏 thanks for the quick reply

**Comment by pedaars** [comment]
I would love to have the ability to automatically tag stories for example if a story has accessibility issues I currently manually mark them with a tag using @Sidnioulz addon-tag-badges like so
 
<img width="780" height="194" alt="Screenshot 2025-10-01 at 14 59 54" src="https://github.com/user-attachments/assets/afd01f3a-d4a7-4f8e-8020-3a78a7f13038" />

But if we could automate this somehow I think that would be great
